[Security Solution][Detection Engine] adds DOES NOT MATCH capability to IM rule type

vitaliidm opened the pull request:
## Summary

 - addresses https://github.com/elastic/security-team/issues/13022
 - allows selecting DOES NOT MATCH condition, it would create alert for any document that has at least one matching field and not matching. Single DOES NOT MATCH is **not allowed**, to prevent creation of large number of alerts. User would be able to select match first and then tune it with DOES NOT MATCH if needed
 - added `negate: boolean` value to ThreatMap entry schema to indicate this mapping should use as DOES NOT MATCH.
 Field name `negate` was chosen as it already used in Kibana filters to indicate field does not match some value.
 - Enrichments for DOES NOT MATCH field are omitted
 - If one of the DOES NOT MATCH fields is empty, we create an alert. If both empty - no alert should be created
  - I fixed issue within `validateCompleteThreatMatches`, where False Positive alert can be created if there partial matches across AND group in multiple threats
  - Removed `x-pack/solutions/security/packages/kbn-securitysolution-io-ts-alerting-types/src/threat_mapping/index.ts` in favour of OpenAPI schema

### Feature Flag

```yml

xpack.securitySolution.enableExperimental:
  - doesNotMatchForIndicatorMatchRuleEnabled
 
```

### UI 

1. Having help text under label and for the first entry, DOES NOT MATCH option should be disabled

<img width="1112" height="214" alt="Screenshot 2025-07-28 at 10 47 26" src="https://github.com/user-attachments/assets/740c8170-45ae-4773-a93b-ac2e0bf12c9c" />

2. User can add DOES NOT MATCH as second AND condition

<img width="1050" height="320" alt="Screenshot 2025-07-28 at 10 47 08" src="https://github.com/user-attachments/assets/56b1f26e-6951-47e7-b7a6-4ff62f5b44a5" />



3. If user deletes first MATCH we would show validation error

<img width="1040" height="241" alt="Screenshot 2025-07-28 at 10 47 43" src="https://github.com/user-attachments/assets/a61274b3-087d-450d-97cc-5ac2b40079db" />

4. user cannot setup a match and not match condition for the same field, validation error

<img width="1064" height="282" alt="Screenshot 2025-07-28 at 10 48 02" src="https://github.com/user-attachments/assets/0ae790e1-7e3d-4fb7-8550-fb3d92b9d678" />

### Flaky test runner

Cypress: https://buildkite.com/elastic/kibana-flaky-test-suite-runner/builds/8886
FTR: https://buildkite.com/elastic/kibana-flaky-test-suite-runner/builds/8885

### Docs issue

- https://github.com/elastic/docs-content/issues/2295

### Test plan

- https://github.com/elastic/security-team/pull/13367

vitaliidm committed:
[Security Solution][Detection Engine] adds NOT MATCHES capability to IM rule type

vitaliidm committed:
[Security Solution][Detection Engine] changes

vitaliidm committed:
[Security Solution][Detection Engine] reanme

vitaliidm committed:
Merge branch 'main' into de_9_2/im-not-matches

vitaliidm committed:
[Security Solution][Detection Engine] adds basic functionality

vitaliidm committed:
[Security Solution][Detection Engine] adjustments

kibanamachine committed:
[CI] Auto-commit changed files from 'node scripts/styled_components_mapping'

elasticmachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/318322)
* Commit: 65c8a0ab27acdb7ab1fc21a994ba118715a21e4a

### Failed CI Steps
* [Quick Checks](https://buildkite.com/elastic/kibana-pull-request/builds/318322#0197fa2e-a93c-4071-856a-96f5e4cbf847)

cc @vitaliidm

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"failing","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/318322","number":318322,"commit":"65c8a0ab27acdb7ab1fc21a994ba118715a21e4a"}],"number":318322}
buildkite-pr-comment-->

kibanamachine committed:
[CI] Auto-commit changed files from 'yarn openapi:bundle'

vitaliidm committed:
Merge branch 'main' into de_9_2/im-not-matches

kibanamachine committed:
[CI] Auto-commit changed files from 'make api-docs'

elasticmachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/318590)
* Commit: b38a11f6bb970d615a9d78c5934ba6c0e110aad9

### Failed CI Steps
* [Check OAS Snapshot](https://buildkite.com/elastic/kibana-pull-request/builds/318590#0198080f-095c-4a48-935b-03916fbc16e9)

### History
* :broken_heart: [Build #318322](https://buildkite.com/elastic/kibana-pull-request/builds/318322) failed 65c8a0ab27acdb7ab1fc21a994ba118715a21e4a

cc @vitaliidm

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"failing","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/318590","number":318590,"commit":"b38a11f6bb970d615a9d78c5934ba6c0e110aad9"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/318322","number":318322,"commit":"65c8a0ab27acdb7ab1fc21a994ba118715a21e4a"}],"number":318590}
buildkite-pr-comment-->

vitaliidm committed:
Merge branch 'main' into de_9_2/im-not-matches

vitaliidm committed:
Merge branch 'main' into de_9_2/im-not-matches

# Conflicts:
#	x-pack/solutions/security/plugins/security_solution/server/lib/detection_engine/rule_types/indicator_match/threat_mapping/build_threat_mapping_filter.ts
#	x-pack/solutions/security/plugins/security_solution/server/lib/detection_engine/rule_types/indicator_match/threat_mapping/get_signals_map_from_threat_index.ts

vitaliidm committed:
[Security Solution][Detection Engine] resolve merge conflicts

vitaliidm committed:
[Security Solution][Detection Engine] add very complicated OR fix for corner case scenario

vitaliidm committed:
[Security Solution][Detection Engine] add tests

vitaliidm committed:
[Security Solution][Detection Engine] fix more tests

vitaliidm committed:
Merge branch 'main' into de_9_2/im-not-matches

vitaliidm committed:
[Security Solution][Detection Engine] fix ine more test

vitaliidm committed:
Merge branch 'main' into de_9_2/im-not-matches

vitaliidm committed:
[Security Solution][Detection Engine]

vitaliidm committed:
[Security Solution][Detection Engine] fix another test

vitaliidm committed:
Merge branch 'main' into de_9_2/im-not-matches

vitaliidm committed:
[Security Solution][Detection Engine] typings

vitaliidm committed:
Merge branch 'de_9_2/im-not-matches' of https://github.com/vitaliidm/kibana into de_9_2/im-not-matches

vitaliidm committed:
[Security Solution][Detection Engine] cleanup

vitaliidm committed:
Merge branch 'main' into de_9_2/im-not-matches

kibanamachine committed:
[CI] Auto-commit changed files from 'yarn openapi:bundle'

elasticmachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/322334)
* Commit: 78993aa5aa3d20fde7c6a516508a0a8a700898a3

### Failed CI Steps
* [Check OAS Snapshot](https://buildkite.com/elastic/kibana-pull-request/builds/322334#019836f9-ede1-443f-b1d7-dc24706e954a)

### History
* :yellow_heart: [Build #321991](https://buildkite.com/elastic/kibana-pull-request/builds/321991) was flaky af9ae5a448b390f584644650ee57f04a9493e246
* :broken_heart: [Build #321923](https://buildkite.com/elastic/kibana-pull-request/builds/321923) failed 72a11e1a7e5f84d035b2c860e27c0b727254e72e
* :broken_heart: [Build #318737](https://buildkite.com/elastic/kibana-pull-request/builds/318737) failed cd9550c5e360daccb54b988dc1427127b2e44d8e
* :broken_heart: [Build #318590](https://buildkite.com/elastic/kibana-pull-request/builds/318590) failed b38a11f6bb970d615a9d78c5934ba6c0e110aad9
* :broken_heart: [Build #318322](https://buildkite.com/elastic/kibana-pull-request/builds/318322) failed 65c8a0ab27acdb7ab1fc21a994ba118715a21e4a

cc @vitaliidm

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"failing","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/322334","number":322334,"commit":"78993aa5aa3d20fde7c6a516508a0a8a700898a3"},{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/321991","number":321991,"commit":"af9ae5a448b390f584644650ee57f04a9493e246"},{"buildStatus":{"state":"failing","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/321923","number":321923,"commit":"72a11e1a7e5f84d035b2c860e27c0b727254e72e"},{"buildStatus":{"state":"failed","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/318737","number":318737,"commit":"cd9550c5e360daccb54b988dc1427127b2e44d8e"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/318590","number":318590,"commit":"b38a11f6bb970d615a9d78c5934ba6c0e110aad9"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/318322","number":318322,"commit":"65c8a0ab27acdb7ab1fc21a994ba118715a21e4a"}],"number":322334}
buildkite-pr-comment-->

vitaliidm committed:
[Security Solution][Detection Engine] fixes and tests

vitaliidm committed:
Merge branch 'de_9_2/im-not-matches' of https://github.com/vitaliidm/kibana into de_9_2/im-not-matches

kibanamachine committed:
[CI] Auto-commit changed files from 'make api-docs'

vitaliidm committed:
Merge branch 'main' into de_9_2/im-not-matches

vitaliidm committed:
[Security Solution][Detection Engine] more tests

vitaliidm committed:
Merge branch 'de_9_2/im-not-matches' of https://github.com/vitaliidm/kibana into de_9_2/im-not-matches

vitaliidm committed:
Merge branch 'main' into de_9_2/im-not-matches

vitaliidm committed:
[Security Solution][Detection Engine] refactoring

vitaliidm committed:
Merge branch 'de_9_2/im-not-matches' of https://github.com/vitaliidm/kibana into de_9_2/im-not-matches

vitaliidm committed:
[Security Solution][Detection Engine] even more tests this time

vitaliidm committed:
[Security Solution][Detection Engine] more tests

elasticmachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/322966)
* Commit: 784f4d5ffc242fc2370d606eb40264ddd448c905

### Failed CI Steps
* [Osquery Cypress Tests on Serverless #8](https://buildkite.com/elastic/kibana-pull-request/builds/322966#01983d24-ed9c-4181-b608-805658584092)

### History
* :broken_heart: [Build #322777](https://buildkite.com/elastic/kibana-pull-request/builds/322777) failed e8ec2b81b73e601c01e205aa84fd26d9f6ff73a2

cc @vitaliidm

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"failing","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/322966","number":322966,"commit":"784f4d5ffc242fc2370d606eb40264ddd448c905"},{"buildStatus":{"state":"failing","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/322777","number":322777,"commit":"e8ec2b81b73e601c01e205aa84fd26d9f6ff73a2"}],"number":322966}
buildkite-pr-comment-->

vitaliidm committed:
Merge branch 'main' into de_9_2/im-not-matches

vitaliidm committed:
[Security Solution][Detection Engine] featre flag

elasticmachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/323274)
* Commit: 7072679c54dfc49d9cb1ca8efacc6006aa5fd626

### Failed CI Steps
* [Osquery Cypress Tests on Serverless #8](https://buildkite.com/elastic/kibana-pull-request/builds/323274#0198414e-2039-4ab7-8308-0c126999c1b7)

### History
* :broken_heart: [Build #323242](https://buildkite.com/elastic/kibana-pull-request/builds/323242) failed ab65444ec48f0bb3f44a3f5563facfcbd5661137
* :broken_heart: [Build #322966](https://buildkite.com/elastic/kibana-pull-request/builds/322966) failed 784f4d5ffc242fc2370d606eb40264ddd448c905
* :broken_heart: [Build #322777](https://buildkite.com/elastic/kibana-pull-request/builds/322777) failed e8ec2b81b73e601c01e205aa84fd26d9f6ff73a2

cc @vitaliidm

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"failing","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/323274","number":323274,"commit":"7072679c54dfc49d9cb1ca8efacc6006aa5fd626"},{"buildStatus":{"state":"failing","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/323242","number":323242,"commit":"ab65444ec48f0bb3f44a3f5563facfcbd5661137"},{"buildStatus":{"state":"failed","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/322966","number":322966,"commit":"784f4d5ffc242fc2370d606eb40264ddd448c905"},{"buildStatus":{"state":"failing","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/322777","number":322777,"commit":"e8ec2b81b73e601c01e205aa84fd26d9f6ff73a2"}],"number":323274}
buildkite-pr-comment-->

vitaliidm committed:
[Security Solution][Detection Engine] fix tests

vitaliidm committed:
[Security Solution][Detection Engine]stylings

vitaliidm committed:
Merge branch 'main' into de_9_2/im-not-matches

elasticmachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/323476)
* Commit: 83da27d64f39d7918a1cee7f31aaccb34c16dab0

### Failed CI Steps
* [Osquery Cypress Tests on Serverless #4](https://buildkite.com/elastic/kibana-pull-request/builds/323476#0198426d-d85e-469f-be2e-554d7f804501)

### History
* :broken_heart: [Build #323367](https://buildkite.com/elastic/kibana-pull-request/builds/323367) failed 07aaf7bf94ad5e5dec2dac9bb84603a70a671d2e
* :broken_heart: [Build #323274](https://buildkite.com/elastic/kibana-pull-request/builds/323274) failed 7072679c54dfc49d9cb1ca8efacc6006aa5fd626
* :broken_heart: [Build #323242](https://buildkite.com/elastic/kibana-pull-request/builds/323242) failed ab65444ec48f0bb3f44a3f5563facfcbd5661137
* :broken_heart: [Build #322966](https://buildkite.com/elastic/kibana-pull-request/builds/322966) failed 784f4d5ffc242fc2370d606eb40264ddd448c905
* :broken_heart: [Build #322777](https://buildkite.com/elastic/kibana-pull-request/builds/322777) failed e8ec2b81b73e601c01e205aa84fd26d9f6ff73a2

cc @vitaliidm

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"failing","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/323476","number":323476,"commit":"83da27d64f39d7918a1cee7f31aaccb34c16dab0"},{"buildStatus":{"state":"failed","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/323367","number":323367,"commit":"07aaf7bf94ad5e5dec2dac9bb84603a70a671d2e"},{"buildStatus":{"state":"failing","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/323274","number":323274,"commit":"7072679c54dfc49d9cb1ca8efacc6006aa5fd626"},{"buildStatus":{"state":"failing","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/323242","number":323242,"commit":"ab65444ec48f0bb3f44a3f5563facfcbd5661137"},{"buildStatus":{"state":"failed","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/322966","number":322966,"commit":"784f4d5ffc242fc2370d606eb40264ddd448c905"},{"buildStatus":{"state":"failing","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/322777","number":322777,"commit":"e8ec2b81b73e601c01e205aa84fd26d9f6ff73a2"}],"number":323476}
buildkite-pr-comment-->

vitaliidm committed:
[Security Solution][Detection Engine] flaky test runner

vitaliidm committed:
Revert "[Security Solution][Detection Engine] flaky test runner"

This reverts commit 12efb3177093586dd4073c070a95634d4d62eec1.

vitaliidm committed:
[Security Solution][Detection Engine] fix test case

elasticmachine commented:
Pinging @elastic/security-detection-engine (Team:Detection Engine)

vitaliidm committed:
Merge branch 'main' into de_9_2/im-not-matches

kibanamachine commented:
## Flaky Test Runner Stats
### 🎉 All tests passed! - [kibana-flaky-test-suite-runner#8886](https://buildkite.com/elastic/kibana-flaky-test-suite-runner/builds/8886)
[✅] Security Solution Detection Engine - Cypress: 100/100 tests passed.

[see run history](https://buildkite.com/elastic/kibana-flaky-test-suite-runner/builds?branch=refs%2Fpull%2F227084%2Fhead)

kibanamachine commented:
## Flaky Test Runner Stats
### 🎉 All tests passed! - [kibana-flaky-test-suite-runner#8885](https://buildkite.com/elastic/kibana-flaky-test-suite-runner/builds/8885)
[✅] x-pack/test/security_solution_api_integration/test_suites/detections_response/detection_engine/rule_execution_logic/indicator_match/trial_license_complete_tier/configs/ess.config.ts: 200/200 tests passed.

[see run history](https://buildkite.com/elastic/kibana-flaky-test-suite-runner/builds?branch=refs%2Fpull%2F227084%2Fhead)

elasticmachine commented:
## :green_heart: Build Succeeded
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/323781)
* Commit: 744c87999f9dd89608401b5dea34a125a54f9ca0

### Metrics [[docs](https://www.elastic.co/guide/en/kibana/master/ci-metrics.html)]
#### Module Count
> Fewer modules leads to a faster build time


| id | [before](https://github.com/elastic/kibana/commit/520c383c0493a8096d3dee47c69b8aeb609fdfcb) | [after](https://github.com/elastic/kibana/commit/744c87999f9dd89608401b5dea34a125a54f9ca0) | diff |
| --- | --- | --- | --- |
| `securitySolution` | 7698 | 7697 | -1 |

#### Public APIs missing comments
> Total count of every public API that lacks a comment. Target amount is 0. Run `node scripts/build_api_docs --plugin [yourplugin] --stats comments` for more detailed information.


| id | [before](https://github.com/elastic/kibana/commit/520c383c0493a8096d3dee47c69b8aeb609fdfcb) | [after](https://github.com/elastic/kibana/commit/744c87999f9dd89608401b5dea34a125a54f9ca0) | diff |
| --- | --- | --- | --- |
| `@kbn/securitysolution-io-ts-alerting-types` | 125 | 87 | -38 |

#### Async chunks
> Total size of all lazy-loaded chunks that will be downloaded as the user navigates the app


| id | [before](https://github.com/elastic/kibana/commit/520c383c0493a8096d3dee47c69b8aeb609fdfcb) | [after](https://github.com/elastic/kibana/commit/744c87999f9dd89608401b5dea34a125a54f9ca0) | diff |
| --- | --- | --- | --- |
| `securitySolution` | 9.9MB | 9.9MB | +1.8KB |

#### Page load bundle
> Size of the bundles that are downloaded on every page load. Target size is below 100kb


| id | [before](https://github.com/elastic/kibana/commit/520c383c0493a8096d3dee47c69b8aeb609fdfcb) | [after](https://github.com/elastic/kibana/commit/744c87999f9dd89608401b5dea34a125a54f9ca0) | diff |
| --- | --- | --- | --- |
| `securitySolution` | 94.3KB | 94.4KB | +44.0B |

<details><summary>Unknown metric groups</summary>

#### API count
| id | [before](https://github.com/elastic/kibana/commit/520c383c0493a8096d3dee47c69b8aeb609fdfcb) | [after](https://github.com/elastic/kibana/commit/744c87999f9dd89608401b5dea34a125a54f9ca0) | diff |
| --- | --- | --- | --- |
| `@kbn/securitysolution-io-ts-alerting-types` | 147 | 109 | -38 |

#### ESLint disabled in files
| id | [before](https://github.com/elastic/kibana/commit/520c383c0493a8096d3dee47c69b8aeb609fdfcb) | [after](https://github.com/elastic/kibana/commit/744c87999f9dd89608401b5dea34a125a54f9ca0) | diff |
| --- | --- | --- | --- |
| `@kbn/securitysolution-io-ts-alerting-types` | 9 | 8 | -1 |

#### Total ESLint disabled count
| id | [before](https://github.com/elastic/kibana/commit/520c383c0493a8096d3dee47c69b8aeb609fdfcb) | [after](https://github.com/elastic/kibana/commit/744c87999f9dd89608401b5dea34a125a54f9ca0) | diff |
| --- | --- | --- | --- |
| `@kbn/securitysolution-io-ts-alerting-types` | 9 | 8 | -1 |


</details>


### History
* :broken_heart: [Build #323476](https://buildkite.com/elastic/kibana-pull-request/builds/323476) failed 83da27d64f39d7918a1cee7f31aaccb34c16dab0
* :broken_heart: [Build #323367](https://buildkite.com/elastic/kibana-pull-request/builds/323367) failed 07aaf7bf94ad5e5dec2dac9bb84603a70a671d2e
* :broken_heart: [Build #323274](https://buildkite.com/elastic/kibana-pull-request/builds/323274) failed 7072679c54dfc49d9cb1ca8efacc6006aa5fd626
* :broken_heart: [Build #323242](https://buildkite.com/elastic/kibana-pull-request/builds/323242) failed ab65444ec48f0bb3f44a3f5563facfcbd5661137
* :broken_heart: [Build #322966](https://buildkite.com/elastic/kibana-pull-request/builds/322966) failed 784f4d5ffc242fc2370d606eb40264ddd448c905

cc @vitaliidm

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/323781","number":323781,"commit":"744c87999f9dd89608401b5dea34a125a54f9ca0"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/323476","number":323476,"commit":"83da27d64f39d7918a1cee7f31aaccb34c16dab0"},{"buildStatus":{"state":"failed","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/323367","number":323367,"commit":"07aaf7bf94ad5e5dec2dac9bb84603a70a671d2e"},{"buildStatus":{"state":"failing","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/323274","number":323274,"commit":"7072679c54dfc49d9cb1ca8efacc6006aa5fd626"},{"buildStatus":{"state":"failing","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/323242","number":323242,"commit":"ab65444ec48f0bb3f44a3f5563facfcbd5661137"},{"buildStatus":{"state":"failed","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/322966","number":322966,"commit":"784f4d5ffc242fc2370d606eb40264ddd448c905"}],"number":323781}
buildkite-pr-comment-->

vitaliidm committed:
Merge branch 'main' into de_9_2/im-not-matches

vitaliidm committed:
Update experimental_features.ts

vitaliidm committed:
Merge branch 'main' into de_9_2/im-not-matches

elasticmachine commented:
## :hourglass_flowing_sand: Build in-progress
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/324193)
* Commit: 86d4584d39eadc790bff16aaf2c544a55489ccba
* [Cloud Deployment](https://kibana-pr-227084.kb.us-west2.gcp.elastic-cloud.com)

### History
* :yellow_heart: [Build #323963](https://buildkite.com/elastic/kibana-pull-request/builds/323963) was flaky 7e123111b5ad276b4be222b39539fc7a9d434b43
* :green_heart: [Build #323781](https://buildkite.com/elastic/kibana-pull-request/builds/323781) succeeded 744c87999f9dd89608401b5dea34a125a54f9ca0
* :broken_heart: [Build #323476](https://buildkite.com/elastic/kibana-pull-request/builds/323476) failed 83da27d64f39d7918a1cee7f31aaccb34c16dab0
* :broken_heart: [Build #323367](https://buildkite.com/elastic/kibana-pull-request/builds/323367) failed 07aaf7bf94ad5e5dec2dac9bb84603a70a671d2e
* :broken_heart: [Build #323274](https://buildkite.com/elastic/kibana-pull-request/builds/323274) failed 7072679c54dfc49d9cb1ca8efacc6006aa5fd626

cc @vitaliidm

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"running","success":true,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/324193","number":324193,"commit":"86d4584d39eadc790bff16aaf2c544a55489ccba"},{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/323963","number":323963,"commit":"7e123111b5ad276b4be222b39539fc7a9d434b43"},{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/323781","number":323781,"commit":"744c87999f9dd89608401b5dea34a125a54f9ca0"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/323476","number":323476,"commit":"83da27d64f39d7918a1cee7f31aaccb34c16dab0"},{"buildStatus":{"state":"failed","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/323367","number":323367,"commit":"07aaf7bf94ad5e5dec2dac9bb84603a70a671d2e"},{"buildStatus":{"state":"failing","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/323274","number":323274,"commit":"7072679c54dfc49d9cb1ca8efacc6006aa5fd626"}],"number":324193}
buildkite-pr-comment-->

nkhristinin commented:
I think I found a potenial bug with enrichments.

I have an IM rule with the following matching conditions:

```
(host.name MATCHES threat.indicator.host.name) AND (user.name DOES NOT MATCH threat.indicator.user.name)
```

#### Event:
- `host.name: host-01`
- `user.name: bob-3`

#### Threat indicators:
1. `threat.indicator.host.name: host-01`, `threat.indicator.user.name: bob-3` — **should not match**
2. `threat.indicator.host.name: host-01`, `threat.indicator.user.name: bob-5` — **should match**

#### Expected:
An alert with a single enrichment from **threat 2** only.

#### Actual:
The alert contains enrichments from both threats, which appears to be incorrect.



<img width="1423" height="766" alt="Screenshot 2025-07-29 at 13 33 42" src="https://github.com/user-attachments/assets/c0832846-8505-4303-996b-4cb2effc742a" />

pmuellr commented:
Given the change in `.../__snapshots__/serverless_upgrade_and_rollback_checks.test.ts.snap`, it seems like a new property was added to an alerting rule type?  `negate`, it appears.

Typically this requires an intermediate release for serverless, since at upgrade time, in a multi-Kibana project, "old" and "new" Kibanas will be running at the same time, so they both need to use the same object shapes.

But I noticed you have a feature flag as well?

> ```yml
> xpack.securitySolution.enableExperimental:
>   - doesNotMatchForIndicatorMatchRuleEnabled
> ```

I assume it's used to start enabling use of this field?  If so, I think we can skip the intermediate release.  I think to hit the verification error you'd see in the old Kibana (doesn't know about the `negate` property), the feature flag would have to be on in the new Kibana, and the field actually used, before the old Kibana was replaced with a new Kibana.  And the rule failure (param verification) would be when the rule runs in the old Kibana.  In theory this COULD be a problem if the old Kibana took a long time to restart, but I think we're probably fine now ...

nkhristinin reviewed (approved):
Reviewed and tested locally, great work!

vitaliidm commented:
@pmuellr 

> Given the change in .../__snapshots__/serverless_upgrade_and_rollback_checks.test.ts.snap, it seems like a new property was added to an alerting rule type? negate, it appears.

This is correct

>But I noticed you have a feature flag as well? I assume it's used to start enabling use of this field?

Yes. It is disabled by default. Typically it takes us 2-3 weeks to enable it on Serverless, until public documentation is ready.


> If so, I think we can skip the intermediate release

Agree with you

vgomez-el commented on x-pack/test/security_solution_cypress/cypress/e2e/detection_response/detection_engine/rule_edit/indicator_match_rule.cy.ts:
```diff
@@ -136,5 +140,60 @@ describe(
       });
     });
+
+    describe('DOES NOT MATCH', () => {
```
@vitaliidm Please, add an appropriate label (@ess, @serverless, etc.) to the top-level describe block so these tests are included in CI runs. Thank you 😄

vgomez-el replied:
My mistake, didn't see the top-level describe. Approving

vgomez-el reviewed (approved):
LGTM!

florent-leborgne commented on x-pack/solutions/security/plugins/security_solution/public/common/components/threat_match/translations.ts:
```diff
@@ -18,6 +18,14 @@ export const THREAT_FIELD = i18n.translate(
+  'xpack.securitySolution.threatMatch.threatFieldLabelHelpDescription',
+  {
+    defaultMessage:
+      'Start with the MATCH condition, and refine it with additional MATCH/DOES NOT MATCH conditions if needed',
```
```suggestion
      'Start with the MATCHES condition, and refine it with additional MATCHES/DOES NOT MATCH conditions if needed',
```
Isn't it "MATCHES"?

florent-leborgne reviewed (approved):
OAS output files LGTM
Left a small comment for the exact wording of "MATCH"/"MATCHES" in the UI copy

vitaliidm committed:
Update x-pack/solutions/security/plugins/security_solution/public/common/components/threat_match/translations.ts

Co-authored-by: florent-leborgne <florent.leborgne@elastic.co>

elasticmachine commented:
## :hourglass_flowing_sand: Build in-progress
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/324743)
* Commit: 6f355e3721515ab85def3604990fc94aca874a16
* [Cloud Deployment](https://kibana-pr-227084.kb.us-west2.gcp.elastic-cloud.com)

### History
* :green_heart: [Build #324193](https://buildkite.com/elastic/kibana-pull-request/builds/324193) succeeded 86d4584d39eadc790bff16aaf2c544a55489ccba
* :yellow_heart: [Build #323963](https://buildkite.com/elastic/kibana-pull-request/builds/323963) was flaky 7e123111b5ad276b4be222b39539fc7a9d434b43
* :green_heart: [Build #323781](https://buildkite.com/elastic/kibana-pull-request/builds/323781) succeeded 744c87999f9dd89608401b5dea34a125a54f9ca0
* :broken_heart: [Build #323476](https://buildkite.com/elastic/kibana-pull-request/builds/323476) failed 83da27d64f39d7918a1cee7f31aaccb34c16dab0
* :broken_heart: [Build #323367](https://buildkite.com/elastic/kibana-pull-request/builds/323367) failed 07aaf7bf94ad5e5dec2dac9bb84603a70a671d2e

cc @vitaliidm

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"running","success":true,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/324743","number":324743,"commit":"6f355e3721515ab85def3604990fc94aca874a16"},{"buildStatus":{"state":"passed","success":true,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/324193","number":324193,"commit":"86d4584d39eadc790bff16aaf2c544a55489ccba"},{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/323963","number":323963,"commit":"7e123111b5ad276b4be222b39539fc7a9d434b43"},{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/323781","number":323781,"commit":"744c87999f9dd89608401b5dea34a125a54f9ca0"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/323476","number":323476,"commit":"83da27d64f39d7918a1cee7f31aaccb34c16dab0"},{"buildStatus":{"state":"failed","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/323367","number":323367,"commit":"07aaf7bf94ad5e5dec2dac9bb84603a70a671d2e"}],"number":324743}
buildkite-pr-comment-->

pmuellr reviewed (approved):
ResponseOps changes LGTM

vitaliidm commented on x-pack/solutions/security/plugins/security_solution/common/api/detection_engine/model/rule_schema/specific_attributes/threat_match_attributes.gen.ts:
```diff
@@ -38,13 +46,7 @@ export type ThreatMapping = z.infer<typeof ThreatMapping>;
-          value: NonEmptyString,
-        })
-      ),
+      entries: z.array(ThreatMapEntry),
```
@marshallmain, In this case we are going to lose order of entries, if someone adds match + negated + match.
To maintain it (probably it's that not important since order does not have impact on rule execution, but I can see how it can be reported as a bug), we probably would need to add addition property to ThreatMapEntry - order. 
Which seems like overkill just for this purpose and because it does not have impact on rule execution.

I was thinking before - throwing error during rule execution, saying some of the rule configuration is not valid. Could only happen when UI validation circumvented and rule update directly by API.

Another option - add manual check in https://github.com/elastic/kibana/blob/main/x-pack/solutions/security/plugins/security_solution/common/api/detection_engine/rule_management/crud/create_rule/request_schema_validation.ts. Likely, the simplest way.

vitaliidm committed:
Merge branch 'main' into de_9_2/im-not-matches

vitaliidm committed:
Merge branch 'main' into de_9_2/im-not-matches

vitaliidm committed:
[Security Solution][Detection Engine] on create validation

vitaliidm committed:
Merge branch 'de_9_2/im-not-matches' of https://github.com/vitaliidm/kibana into de_9_2/im-not-matches

elasticmachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/325058)
* Commit: 9d4ca0a885309b6034eb8c00c3476c0f93806a9a

### Failed CI Steps
* [Quick Checks](https://buildkite.com/elastic/kibana-pull-request/builds/325058#01986031-93d9-441c-8cd9-b62830c4d72c)

### History
* :broken_heart: [Build #325041](https://buildkite.com/elastic/kibana-pull-request/builds/325041) failed aa605c0e82fce73ed5c1a23c9061f7dbb7c12f57
* :broken_heart: [Build #325015](https://buildkite.com/elastic/kibana-pull-request/builds/325015) failed 1a02d91f019abec0e0b0b32892d909677e927096
* :broken_heart: [Build #324743](https://buildkite.com/elastic/kibana-pull-request/builds/324743) failed 6f355e3721515ab85def3604990fc94aca874a16
* :green_heart: [Build #324193](https://buildkite.com/elastic/kibana-pull-request/builds/324193) succeeded 86d4584d39eadc790bff16aaf2c544a55489ccba
* :yellow_heart: [Build #323963](https://buildkite.com/elastic/kibana-pull-request/builds/323963) was flaky 7e123111b5ad276b4be222b39539fc7a9d434b43
* :green_heart: [Build #323781](https://buildkite.com/elastic/kibana-pull-request/builds/323781) succeeded 744c87999f9dd89608401b5dea34a125a54f9ca0

cc @vitaliidm

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"failing","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/325058","number":325058,"commit":"9d4ca0a885309b6034eb8c00c3476c0f93806a9a"},{"buildStatus":{"state":"failing","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/325041","number":325041,"commit":"aa605c0e82fce73ed5c1a23c9061f7dbb7c12f57"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/325015","number":325015,"commit":"1a02d91f019abec0e0b0b32892d909677e927096"},{"buildStatus":{"state":"failed","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/324743","number":324743,"commit":"6f355e3721515ab85def3604990fc94aca874a16"},{"buildStatus":{"state":"passed","success":true,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/324193","number":324193,"commit":"86d4584d39eadc790bff16aaf2c544a55489ccba"},{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/323963","number":323963,"commit":"7e123111b5ad276b4be222b39539fc7a9d434b43"}],"number":325058}
buildkite-pr-comment-->

vitaliidm committed:
Merge branch 'main' into de_9_2/im-not-matches

elasticmachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/325095)
* Commit: dfd103d70dd6b2f2c847d03195a3a07a90f8edd4

### Failed CI Steps
* [Quick Checks](https://buildkite.com/elastic/kibana-pull-request/builds/325095#0198606a-6fce-4c2d-8988-95719138632d)

### History
* :broken_heart: [Build #325058](https://buildkite.com/elastic/kibana-pull-request/builds/325058) failed 9d4ca0a885309b6034eb8c00c3476c0f93806a9a
* :broken_heart: [Build #325041](https://buildkite.com/elastic/kibana-pull-request/builds/325041) failed aa605c0e82fce73ed5c1a23c9061f7dbb7c12f57
* :broken_heart: [Build #325015](https://buildkite.com/elastic/kibana-pull-request/builds/325015) failed 1a02d91f019abec0e0b0b32892d909677e927096
* :broken_heart: [Build #324743](https://buildkite.com/elastic/kibana-pull-request/builds/324743) failed 6f355e3721515ab85def3604990fc94aca874a16
* :green_heart: [Build #324193](https://buildkite.com/elastic/kibana-pull-request/builds/324193) succeeded 86d4584d39eadc790bff16aaf2c544a55489ccba

cc @vitaliidm

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"failing","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/325095","number":325095,"commit":"dfd103d70dd6b2f2c847d03195a3a07a90f8edd4"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/325058","number":325058,"commit":"9d4ca0a885309b6034eb8c00c3476c0f93806a9a"},{"buildStatus":{"state":"failing","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/325041","number":325041,"commit":"aa605c0e82fce73ed5c1a23c9061f7dbb7c12f57"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/325015","number":325015,"commit":"1a02d91f019abec0e0b0b32892d909677e927096"},{"buildStatus":{"state":"failed","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/324743","number":324743,"commit":"6f355e3721515ab85def3604990fc94aca874a16"},{"buildStatus":{"state":"passed","success":true,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/324193","number":324193,"commit":"86d4584d39eadc790bff16aaf2c544a55489ccba"}],"number":325095}
buildkite-pr-comment-->

vitaliidm committed:
Merge branch 'main' into de_9_2/im-not-matches

nikitaindik commented on x-pack/solutions/security/plugins/security_solution/common/api/detection_engine/rule_management/crud/patch_rule/request_schema_validation.ts:
```diff
@@ -17,6 +18,7 @@ export const validatePatchRuleRequestBody = (rule: PatchRuleRequestBody): string
     ...validateTimelineId(rule),
     ...validateTimelineTitle(rule),
     ...validateThreshold(rule),
+    ...validateThreatMapping(rule),
```
Is see that `validateThreatMatchConcurrentSearches` is used for the creation. Do we want to use it for patch/update as well?

vitaliidm replied:
I have created a ticket for that since it's already a case in main and out of scope for this feature: https://github.com/elastic/kibana/issues/230251

vitaliidm committed:
Merge branch 'main' into de_9_2/im-not-matches

# Conflicts:
#	x-pack/test/security_solution_cypress/cypress/e2e/detection_response/rule_management/prebuilt_rules/prebuilt_rules_preview.cy.ts

vitaliidm committed:
[Security Solution][Detection Engine] fix conflicts

vitaliidm committed:
[Security Solution][Detection Engine] fix lint

kibanamachine committed:
[CI] Auto-commit changed files from 'node scripts/eslint_all_files --no-cache --fix'

marshallmain commented:
I ran out of time to review this again this week but will review on Monday 👀

nikitaindik reviewed (approved):
Thanks for addressing my comments, @vitaliidm.

vitaliidm committed:
Merge branch 'main' into de_9_2/im-not-matches

vitaliidm committed:
[Security Solution][Detection Engine]

vitaliidm committed:
Merge branch 'de_9_2/im-not-matches' of https://github.com/vitaliidm/kibana into de_9_2/im-not-matches

elasticmachine commented:
## :broken_heart: Build Failed
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/325992)
* Commit: 4b5a4d27a7628a3ef00735322b5a691ec12b5d0b

### Failed CI Steps
* [Check Types](https://buildkite.com/elastic/kibana-pull-request/builds/325992#019874f1-1e01-4081-a49e-171dff69df91)

### History
* :broken_heart: [Build #325680](https://buildkite.com/elastic/kibana-pull-request/builds/325680) failed bd69b5c6aafc6da3ab3ac1a0b5959017b7a37e05
* :broken_heart: [Build #325565](https://buildkite.com/elastic/kibana-pull-request/builds/325565) failed 9f6e30e4d8f2d8e41412327210b9ab6d9d2195c2
* :green_heart: [Build #325152](https://buildkite.com/elastic/kibana-pull-request/builds/325152) succeeded 56300b356c9e58cb7a8748f7f33c4f00a9aa5403
* :broken_heart: [Build #325095](https://buildkite.com/elastic/kibana-pull-request/builds/325095) failed dfd103d70dd6b2f2c847d03195a3a07a90f8edd4
* :broken_heart: [Build #325058](https://buildkite.com/elastic/kibana-pull-request/builds/325058) failed 9d4ca0a885309b6034eb8c00c3476c0f93806a9a
* :broken_heart: [Build #325041](https://buildkite.com/elastic/kibana-pull-request/builds/325041) failed aa605c0e82fce73ed5c1a23c9061f7dbb7c12f57

cc @vitaliidm

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/325992","number":325992,"commit":"4b5a4d27a7628a3ef00735322b5a691ec12b5d0b"},{"buildStatus":{"state":"failed","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/325680","number":325680,"commit":"bd69b5c6aafc6da3ab3ac1a0b5959017b7a37e05"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/325565","number":325565,"commit":"9f6e30e4d8f2d8e41412327210b9ab6d9d2195c2"},{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/325152","number":325152,"commit":"56300b356c9e58cb7a8748f7f33c4f00a9aa5403"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/325095","number":325095,"commit":"dfd103d70dd6b2f2c847d03195a3a07a90f8edd4"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/325058","number":325058,"commit":"9d4ca0a885309b6034eb8c00c3476c0f93806a9a"}],"number":325992}
buildkite-pr-comment-->

vitaliidm committed:
[Security Solution][Detection Engine] fix lint

vitaliidm committed:
Merge branch 'main' into de_9_2/im-not-matches

marshallmain commented:
Mostly minor things, but two bug fixes I think are important:
- [Concatenating field names together to create unique keys](https://github.com/elastic/kibana/pull/227084#discussion_r2251989606)
- [Handling clause generation failures](https://github.com/elastic/kibana/pull/227084#discussion_r2252398149)

vitaliidm committed:
[Security Solution][Detection Engine] rename ThreatMapEntry

vitaliidm committed:
[Security Solution][Detection Engine] flag

kibanamachine committed:
[CI] Auto-commit changed files from 'yarn openapi:bundle'

kibanamachine committed:
[CI] Auto-commit changed files from 'make api-docs'

vitaliidm committed:
Merge branch 'main' into de_9_2/im-not-matches

vitaliidm committed:
[Security Solution][Detection Engine] remove partial entries

vitaliidm committed:
[Security Solution][Detection Engine] typings

vitaliidm committed:
[Security Solution][Detection Engine]
      - const threatListItem = getThreatListSearchResponseMock().hits.hits[0];
      +const threatListItem = getThreatListItemMock();

vitaliidm committed:
[Security Solution][Detection Engine] negateMap

vitaliidm commented on x-pack/solutions/security/plugins/security_solution/common/utils/request_validation/indicator_match.ts:
```diff
@@ -0,0 +1,88 @@
+    const negateSet = new Set(
+      item.entries
+        .filter(({ negate, field, value }) => negate && field && value)
+        .map((subEntry) => `${subEntry.field}-${subEntry.value}`)
```
refactored function

vitaliidm commented on x-pack/solutions/security/plugins/security_solution/public/common/components/threat_match/translations.ts:
```diff
@@ -18,6 +18,14 @@ export const THREAT_FIELD = i18n.translate(
+  'xpack.securitySolution.threatMatch.threatFieldLabelHelpDescription',
+  {
+    defaultMessage:
+      'Start by defining the MATCHES conditions, then add more MATCHES/DOES NOT MATCH entries as needed.',
```
This edit was suggested by @nastasha-solomon , since user can add multiple MATCHES conditions before adding DOES NOT MATCH.

vitaliidm commented on x-pack/solutions/security/plugins/security_solution/public/detection_engine/rule_management/components/rule_details/rule_definition_section.tsx:
```diff
@@ -386,17 +384,21 @@ interface ThreatMappingProps {
 
-export const ThreatMapping = ({ threatMapping }: ThreatMappingProps) => {
-  const description = threatMapping.reduce<string>(
+export const constructThreatMappingDescription = (threatMapping: ThreatMappingType): string => {
```
I am not sure about this change. This will replace current representation of threat mappings across all existing rules with more the than one entry. 
This function was extracted from existing components(where it was duplicated) and does not change any existing behaviour.

If we do want to make this change, I think it should be a separate issue with possible involvement of doc team

marshallmain replied:
Sounds good, we can look into this later on 👍

vitaliidm committed:
[Security Solution][Detection Engine] invert loops

vitaliidm committed:
Update x-pack/solutions/security/plugins/security_solution/server/lib/detection_engine/rule_types/indicator_match/threat_mapping/utils.test.ts

Co-authored-by: Marshall Main <55718608+marshallmain@users.noreply.github.com>

kibanamachine committed:
[CI] Auto-commit changed files from 'node scripts/eslint_all_files --no-cache --fix'

vitaliidm committed:
Merge branch 'main' into de_9_2/im-not-matches

vitaliidm committed:
Update x-pack/solutions/security/plugins/security_solution/public/common/components/threat_match/translations.ts

Co-authored-by: Marshall Main <55718608+marshallmain@users.noreply.github.com>

nastasha-solomon commented on x-pack/solutions/security/plugins/security_solution/public/detection_engine/rule_creation/components/threat_match_mapping_edit/validators/threat_match_mapping_validator_factory.ts:
```diff
@@ -112,5 +116,32 @@ export function threatMatchMappingValidatorFactory({
+        message: i18n.translate(
+          'xpack.securitySolution.detectionEngine.ruleManagement.threatMappingField.singleNotMatchClauseError',
+          {
+            defaultMessage: 'Conditions in AND clauses must have at least one MATCHES entry.',
```
```suggestion
            defaultMessage: 'Conditions with AND clauses must have at least one MATCHES entry.',
```

nastasha-solomon reviewed (approved):
Just the one small tweak. Otherwise, lgtm for now!

vitaliidm committed:
[Security Solution][Detection Engine] wording

vitaliidm committed:
Merge branch 'de_9_2/im-not-matches' of https://github.com/vitaliidm/kibana into de_9_2/im-not-matches

vitaliidm committed:
Merge branch 'main' into de_9_2/im-not-matches

marshallmain commented on x-pack/solutions/security/plugins/security_solution/server/lib/detection_engine/rule_types/indicator_match/threat_mapping/validate_complete_threat_matches.ts:
```diff
@@ -66,28 +66,42 @@ export const validateCompleteThreatMatches = (
-        return false;
-      });
+    Object.values(threatQueriesMap).forEach((threatQueriesPerId) => {
+      const matchedThreatQueriesForAndGroup: ThreatMatchNamedQuery[] = [];
```
This `const` can go inside the inner loop as well?

marshallmain reviewed (approved):
LGTM, thanks for addressing my comments 🚀 

Understanding the changes to `validateCompleteThreatMatches` was by far the hardest part, and I hope we can drastically simplify that logic by adjusting the way we name query clauses. The remaining unresolved comments are for future (soon) work.

vitaliidm committed:
[Security Solution][Detection Engine] loop

vitaliidm committed:
Merge branch 'de_9_2/im-not-matches' of https://github.com/vitaliidm/kibana into de_9_2/im-not-matches

elasticmachine commented:
## :hourglass_flowing_sand: Build in-progress
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/326826)
* Commit: e6d68c9afb7caaa484711b3a9d15b06ce1dac002
* [Cloud Deployment](https://kibana-pr-227084.kb.us-west2.gcp.elastic-cloud.com)

### History
* :green_heart: [Build #326520](https://buildkite.com/elastic/kibana-pull-request/builds/326520) succeeded e66f4860eca303e662f291a068875042e753b28e
* :broken_heart: [Build #326432](https://buildkite.com/elastic/kibana-pull-request/builds/326432) failed 52b79dad13684d0b1895f46e6509e877e6096585
* :green_heart: [Build #326026](https://buildkite.com/elastic/kibana-pull-request/builds/326026) succeeded bbdd39a3d29077d1062501f0930b0f44ff89a21a
* :broken_heart: [Build #325992](https://buildkite.com/elastic/kibana-pull-request/builds/325992) failed 4b5a4d27a7628a3ef00735322b5a691ec12b5d0b
* :broken_heart: [Build #325680](https://buildkite.com/elastic/kibana-pull-request/builds/325680) failed bd69b5c6aafc6da3ab3ac1a0b5959017b7a37e05
* :broken_heart: [Build #325565](https://buildkite.com/elastic/kibana-pull-request/builds/325565) failed 9f6e30e4d8f2d8e41412327210b9ab6d9d2195c2

cc @vitaliidm

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"running","success":true,"hasRetries":true,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/326826","number":326826,"commit":"e6d68c9afb7caaa484711b3a9d15b06ce1dac002"},{"buildStatus":{"state":"running","success":true,"hasRetries":true,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/326520","number":326520,"commit":"e66f4860eca303e662f291a068875042e753b28e"},{"buildStatus":{"state":"failing","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/326432","number":326432,"commit":"52b79dad13684d0b1895f46e6509e877e6096585"},{"buildStatus":{"state":"running","success":true,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/326026","number":326026,"commit":"bbdd39a3d29077d1062501f0930b0f44ff89a21a"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/325992","number":325992,"commit":"4b5a4d27a7628a3ef00735322b5a691ec12b5d0b"},{"buildStatus":{"state":"failed","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/325680","number":325680,"commit":"bd69b5c6aafc6da3ab3ac1a0b5959017b7a37e05"}],"number":326826}
buildkite-pr-comment-->

vitaliidm committed:
Merge branch 'main' into de_9_2/im-not-matches

elasticmachine commented:
## :hourglass_flowing_sand: Build in-progress
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/326902)
* Commit: 17b73718253a715340d72c4f5ff4537539a83dde
* [Cloud Deployment](https://kibana-pr-227084.kb.us-west2.gcp.elastic-cloud.com)

### History
* :broken_heart: [Build #326826](https://buildkite.com/elastic/kibana-pull-request/builds/326826) failed e6d68c9afb7caaa484711b3a9d15b06ce1dac002
* :green_heart: [Build #326520](https://buildkite.com/elastic/kibana-pull-request/builds/326520) succeeded e66f4860eca303e662f291a068875042e753b28e
* :broken_heart: [Build #326432](https://buildkite.com/elastic/kibana-pull-request/builds/326432) failed 52b79dad13684d0b1895f46e6509e877e6096585
* :green_heart: [Build #326026](https://buildkite.com/elastic/kibana-pull-request/builds/326026) succeeded bbdd39a3d29077d1062501f0930b0f44ff89a21a
* :broken_heart: [Build #325992](https://buildkite.com/elastic/kibana-pull-request/builds/325992) failed 4b5a4d27a7628a3ef00735322b5a691ec12b5d0b

cc @vitaliidm

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"running","success":true,"hasRetries":true,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/326902","number":326902,"commit":"17b73718253a715340d72c4f5ff4537539a83dde"},{"buildStatus":{"state":"failing","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/326826","number":326826,"commit":"e6d68c9afb7caaa484711b3a9d15b06ce1dac002"},{"buildStatus":{"state":"running","success":true,"hasRetries":true,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/326520","number":326520,"commit":"e66f4860eca303e662f291a068875042e753b28e"},{"buildStatus":{"state":"failing","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/326432","number":326432,"commit":"52b79dad13684d0b1895f46e6509e877e6096585"},{"buildStatus":{"state":"running","success":true,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/326026","number":326026,"commit":"bbdd39a3d29077d1062501f0930b0f44ff89a21a"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/325992","number":325992,"commit":"4b5a4d27a7628a3ef00735322b5a691ec12b5d0b"}],"number":326902}
buildkite-pr-comment-->
